tasks.py

```diff
@@ -0,0 +1,261 @@
+            print(f"Deploying to {space} due to matching branch name {branch}")
+            return space
+    print(f"Current branch {branch} does not match any deployment specifications.")
+    print(f"Skipping deployment.")
+    return None
+
+def _resolve_environment_file(repo, branch):
+    """Get spacenvironment file associated with first matching rule."""
+    for file, rule in ENVIRONMENT_FILE_RULES:
+        if rule(repo, branch):
+            print(f"Using env file {file} due to matching branch name {branch}")
+            return file
+    print(f"Current branch {branch} does not match any deployment specifications.")
+    return None
+
+
+def _detect_branch(repo):
+    try:
+        return repo.active_branch.name
+    except TypeError:
+        return None
+
+
+DEPLOY_RULES = (
+    ("prod", _detect_prod),
+    ("stage", lambda _, branch: branch.startswith("release")),
+    ("dev", lambda _, branch: branch == "develop"),
+    #("dev", lambda _, branch: branch == "feature/40-deploy-to-cloud.gov"),
+)
+
+def _build_angular_app(ctx,space):
+    orig_directory = os.getcwd()
+    os.chdir(os.path.join(orig_directory, 'front-end'))
+
+    ctx.run("npm install", warn=True, echo=True)
+    print(f"Starting build: npm run build-{space}")
+    result = ctx.run(f"npm run build-{space}", warn=True, echo=True)
+
+    if result.return_code != 0:
+        print (f"error building Angular app.  Exiting with code {result.return_code}")
+        exit(result.return_code)
+
+    os.chdir (orig_directory)
+
+# copies a few nginx config files into the Angualr app distribution directory
+def _prep_distribution_directory(ctx):
+    dist_directory = os.path.join(os.getcwd(), 'front-end', 'dist')
+    nginx_config_dir = os.path.join(os.getcwd(), 'deploy-config', 'front-end-nginx-config')
+
+    copyfile(os.path.join(nginx_config_dir, 'nginx.conf'), os.path.join(dist_directory, 'nginx.conf'))
+    copyfile(os.path.join(nginx_config_dir, 'mime.types'), os.path.join(dist_directory, 'mime.types'))
+
+
+def _login_to_cf(ctx, space):
+    # Set api
+    api = "https://api.fr.cloud.gov"
+    ctx.run(f"cf api {api}", echo=True)
+    # Authenticate
+    login_command = 'cf auth "$FEC_CF_USERNAME_{0}" "$FEC_CF_PASSWORD_{0}"'.format(
+        space.upper()
+    )
+    result = ctx.run(login_command, echo=True, warn=True)
+    if result.return_code != 0:
+        print("\n\nError logging into cloud.gov.")
+        print("Please check your authentication environment variables:")
+
+        user_var = f'$FEC_CF_USERNAME_{space.upper()}'
+        pass_var = f'$FEC_CF_PASSWORD_{space.upper()}'
+
+        if os.getenv(user_var) and os.getenv(pass_var):
+            user = os.getenv(user_var)
+            passwd = os.getenv(pass_var) or ''
+            print(f"passwd is {passwd}")
+            safe_passwd = ('*' * len(passwd )) + os.getenv(passwd)[-3:]
+
+            print(f"${user_var} = {user}");
+            print(f"${pass_var} = {safe_passwd}");
```

Comment:
Even though we're only printing user and a partial password, this gives me pause. This gets output to CircleCI, which is public. Would you be open to remove printing out the user and safe password?

Reply:
I'm fine removing it.  The purpose is to help with debugging later. I'll print the env var names only and that should be enough to help people out.